File: examples/example_correctionlib_electron.py
#!/usr/bin/env python3
from pathlib import Path
from types import SimpleNamespace

import correctionlib


repo = Path(__file__).resolve().parents[1]
payload = (
    repo
    / "payloads/Run3_NanoAODv15_correctionlib/EGM/Electron/eff_classifier_pMSSM_334_105076_GenElectron_loose.correctionlib.json"
)
corr = correctionlib.CorrectionSet.from_file(str(payload))[
    "fastsim_sf_pMSSM_334_105076_GenElectron_loose"
]
tiny_nano = repo / "examples/data/ttbar_fastsim_nano_3events.root"


def fastsim2fullsim_sf_from_kinematics(pt, eta, phi, iso):
    return corr.evaluate(pt, eta, phi, iso)


def bind_fastsim2fullsim_sf_to_event(event):
    def fastsim2fullsim_sf_from_gen_index(gen_index):
        if gen_index < 0:
            return 1.0
        return fastsim2fullsim_sf_from_kinematics(
            event.GenPart_pt[gen_index],
            event.GenPart_eta[gen_index],
            event.GenPart_phi[gen_index],
            event.GenPart_iso[gen_index],
        )

    return fastsim2fullsim_sf_from_gen_index


def tiny_nano_events():
    try:
        import ROOT
    except ImportError:
        yield SimpleNamespace(
            Electron_genPartIdx=[2],
            GenPart_pt=[5.0, 12.0, 15.0],
            GenPart_eta=[-1.1, 1.3, 0.4],
            GenPart_phi=[0.2, -2.4, 2.1],
            GenPart_iso=[0.02, 0.01, 1e-3],
        )
        return

    nano_file = ROOT.TFile.Open(str(tiny_nano))
    events = nano_file.Get("Events")
    for event in events:
        yield event


for event_index, event in enumerate(tiny_nano_events()):
    sf_from_gen_index = bind_fastsim2fullsim_sf_to_event(event)
    # Analyst chooses collection, applies ID/isolation/selection cuts, then gets the SF.
    for electron_index, gen_index in enumerate(event.Electron_genPartIdx):
        sf = sf_from_gen_index(gen_index)
        print(f"event {event_index} electron {electron_index}: sf_full_over_fast {sf:.17g}")

sf_scan = fastsim2fullsim_sf_from_kinematics(pt=15.0, eta=0.4, phi=2.1, iso=1e-3)
print(f"kinematic scan point: sf_full_over_fast {sf_scan:.17g}")

File: examples/example_correctionlib_electron.cc
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

#include "TFile.h"
#include "TTreeReader.h"
#include "TTreeReaderArray.h"
#include "correction.h"

double sf_from_kinematics(const correction::Correction::Ref& corr, double pt, double eta, double phi, double iso) {
  return corr->evaluate(std::vector<correction::Variable::Type>{pt, eta, phi, iso});
}

double sf_from_gen_index(const correction::Correction::Ref& corr, int gen_index, const TTreeReaderArray<float>& GenPart_pt, const TTreeReaderArray<float>& GenPart_eta, const TTreeReaderArray<float>& GenPart_phi, const TTreeReaderArray<float>& GenPart_iso) {
  if (gen_index < 0 || gen_index >= GenPart_pt.GetSize()) return 1.0;
  return sf_from_kinematics(corr, GenPart_pt[gen_index], GenPart_eta[gen_index], GenPart_phi[gen_index], GenPart_iso[gen_index]);
}

int main() {
  const std::string payload = "payloads/Run3_NanoAODv15_correctionlib/EGM/Electron/eff_classifier_pMSSM_334_105076_GenElectron_loose.correctionlib.json";
  const std::string correction_name = "fastsim_sf_pMSSM_334_105076_GenElectron_loose";
  const std::string tiny_nano = "examples/data/ttbar_fastsim_nano_3events.root";

  auto cset = correction::CorrectionSet::from_file(payload);
  auto corr = cset->at(correction_name);

  TFile nano_file(tiny_nano.c_str());
  TTreeReader events("Events", &nano_file);
  TTreeReaderArray<short> Electron_genPartIdx(events, "Electron_genPartIdx");
  TTreeReaderArray<float> GenPart_pt(events, "GenPart_pt");
  TTreeReaderArray<float> GenPart_eta(events, "GenPart_eta");
  TTreeReaderArray<float> GenPart_phi(events, "GenPart_phi");
  TTreeReaderArray<float> GenPart_iso(events, "GenPart_iso");

  std::cout << std::setprecision(17);
  int event_index = 0;
  while (events.Next()) {
    // Analyst chooses collection, applies ID/isolation/selection cuts, then gets the SF.
    for (int electron_index = 0; electron_index < Electron_genPartIdx.GetSize(); ++electron_index) {
      const int gen_index = Electron_genPartIdx[electron_index];
      const double sf = sf_from_gen_index(corr, gen_index, GenPart_pt, GenPart_eta, GenPart_phi, GenPart_iso);
      std::cout << "event " << event_index << " electron " << electron_index << ": sf_full_over_fast " << sf << "\n";
    }
    ++event_index;
  }

  const double sf_scan = sf_from_kinematics(corr, 15.0, 0.4, 2.1, 1e-3);
  std::cout << "kinematic scan point: sf_full_over_fast " << sf_scan << "\n";
  return 0;
}
